### ISSUE TITLE
Exec daemon unexpected exit

### ISSUE THREAD

**Issue by chemamartinez** [wazuh/wazuh#3227]
|Wazuh version|Install type|Install method|Platform|
|---|---|---|---|
| 3.8.2 -> 3.9.0 | Agent | WPK upgrade | OpenSuse 42.3 |

The `execd` daemon crashes when the WPK upgrade procedure starts at the same time the agent is receiving the shared configuration. It has been reproduced upgrading by WPK from Wazuh v3.8.2 to v3.9.0.

Here we can see the implies logs:

- Agent

```
2019/05/02 10:42:42 ossec-agentd: INFO: Agent is restarting due to shared configuration changes.
2019/05/02 10:42:42 ossec-execd: INFO: Agent auto-restart locked for 541 seconds.
2019/05/02 10:42:44 ossec-agentd: ERROR: At req_push(): Target 'com' refused connection. The component might be disabled
2019/05/02 10:42:45 ossec-syscheckd: INFO: Ending syscheck scan. Database completed.
2019/05/02 10:42:48 ossec-agentd: ERROR: At req_push(): Target 'com' refused connection. The component might be disabled
2019/05/02 10:42:53 ossec-agentd: ERROR: At req_push(): Target 'com' refused connection. The component might be disabled
2019/05/02 10:42:57 ossec-agentd: ERROR: At req_push(): Target 'com' refused connection. The component might be disabled
2019/05/02 10:43:02 ossec-agentd: ERROR: At req_push(): Target 'com' refused connection. The component might be disabled
2019/05/02 10:43:06 ossec-agentd: ERROR: At req_push(): Target 'com' refused connection. The component might be disabled
2019/05/02 10:43:11 ossec-agentd: ERROR: At req_push(): Target 'com' refused connection. The component might be disabled
2019/05/02 10:43:12 ossec-agentd: INFO: Agent is restarting due to shared configuration changes.
2019/05/02 10:43:12 ossec-agentd: ERROR: Could not auto-restart agent. Is Active Response enabled?
```

- Manager:

```
2019/05/02 10:27:53 ossec-remoted: WARNING: At req_dispatch(): OS_SendSecureTCP(): Broken pipe
```

Finally, the command output to trigger the upgrade when the agent is restarting because of the shared configuration:

```
# agent_upgrade -a 011 -d -r <wpk-repository>
Manager version: v3.9.0
Agent version: v3.8.2
Agent new version: v3.9.0
WPK file already downloaded: /var/ossec/var/upgrade/wazuh_agent_v3.9.0_linux_x86_64.wpk - SHA1SUM: 180e25a1fefafe8d83c763d375cb1a3a387bc08a
Upgrade PKG: wazuh_agent_v3.9.0_linux_x86_64.wpk (4492.53125 KB)
MSG SENT: 011 com open wb wazuh_agent_v3.9.0_linux_x86_64.wpk
RESPONSE: err Connection refused
```

After the connection refused message, the exec daemon is down on the agent side.

**Comment by cristgl** [comment]
This is the output when sending the wpk to the agent:
```
2019/05/28 15:01:41 ossec-remoted[573] request.c:239 at req_dispatch(): DEBUG: Sending request: '#!-req bd160219 com write 512 wazuh_agent_v3.9.1_linux_x86_64.wpk ��`������?
                                                                           �
                                                                              �1�c0��`�����sWM��qO���S���X�ZcE�����5��yd�7N��������`��@�^:�X{��?liQx���t��kk�C'ͧHr��[t��@fbc�1T�
                                                                                    &�����`�0�B��Z-��yV(���7o�����k�T������?��
                             4�2m4�r�;ga|j��=�b��^N/�;��"T�vK�,�=����ca�"<h�6��`h�A��
                                                                                        �
Ce���m��ŷOX����%�'��_ځ�WA�{'
```

And at the agent side:
```
2019/05/28 16:33:16 ossec-agentd[20416] receiver.c:84 at receive_msg(): DEBUG: Received message: '#!-req 427fcb14 com open wb wazuh_agent_v3.9.1_linux_x86_64.wpk'
2019/05/28 16:33:16 ossec-agentd[20416] request.c:152 at req_push(): ERROR: At req_push(): Target 'com' refused connection. The component might be disabled
2019/05/28 16:33:16 ossec-agentd[20416] receiver.c:84 at receive_msg(): DEBUG: Received message: '#!-req 427fcb14 ack'
2019/05/28 16:33:16 ossec-agentd[20416] request.c:129 at req_push(): DEBUG: Request counter (427fcb14) not found. Duplicated ACK?
```

**Comment by cristgl** [comment]
Observed behavior:
- The error occurs only if the `auto-restart` is locked: 
```
2019/05/30 13:43:20 ossec-execd: INFO: Agent auto-restart locked for 540 seconds.
2019/05/30 13:43:21 ossec-agentd: ERROR: At req_push(): Target 'com' refused connection. The component might be disabled
```
- If it is not locked, the output is `No file is opened`. This is related to the `wcom_write` function, which may have not a file path in which writing.

- On Ubuntu, I have find out that, when `auto-restart` is locked, the file is closed:
`2019/05/30 10:25:41 ossec-execd: ERROR: File '/var/ossec//var/incoming/wazuh_agent_v3.9.1_linux_x86_64.wpk' was opened. Closing.`
Which should happen on SUSE, too, but somehow the timeout for locking `auto-restart` is overtaken.

**Comment by cristgl** [comment]
**Bug when `auto-restart` is not locked**

This bug occurs when upgrading the agent while auto-restarting, but when this one is locked. This means that it will pass by this [`else`](https://github.com/wazuh/wazuh/blob/80b09587120ab673e3f7982c8aac8e5aac412359/src/os_execd/wcom.c#L556) statement.

When it is not locked, the output is the next one.
```
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
2019/05/31 12:15:29 ossec-execd: ERROR: At WCOM write: No file is opened.
``` 

Which means that this file was going to be written but it is not opened.

**Comment by cristgl** [comment]
Continuing with the bug when auto-restart is not locked, the solution followed was the next one:

- Add a more descriptive message when the agent does not have the file opened:
From `No file opened` to `File not opened. Agent might have been auto-restarted during upgrade`.
- Stop sending the WPK at the manager side if the received message from the agent was an error.
- Exchange the `open` and `lock_restart` messages sent to the agent so that the file is not opened until the `auto-restart` is locked.

**Comment by chemamartinez** [comment]
Merged: #3437
